### PR TITLE
[Track2] Add impl of service plan and deployment slot manager

### PR DESCRIPTION
**Pull request by Flanker32**
Add impl of service plan and deployment slot manager

### CODE REVIEW

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkits/appservice/service/impl/AppServicePlan.java [hunk 1]
```diff
@@ -30,36 +36,132 @@ public AppServicePlan(AppServicePlanEntity appServicePlanEntity, AzureAppService
...
 
     @Override
     public String name() {
-        return null;
+        return getPlanService().name();
+    }
+
+    public class AppServicePlanCreator implements IAppServicePlanCreator {
+        private String name;
+        private Region region;
```
Comment (andxu) [inline azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkits/appservice/service/impl/AppServicePlan.java:88]: shall we share the common fields in one base class: the name region, resource group?

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkits/appservice/service/impl/AppServicePlan.java [hunk 2]
```diff
@@ -30,36 +36,132 @@ public AppServicePlan(AppServicePlanEntity appServicePlanEntity, AzureAppService
...
+            return this;
+        }
+
+        @Override
+        public AppServicePlan commit() {
+            boolean modified = false;
+            com.azure.resourcemanager.appservice.models.AppServicePlan.Update update = appServicePlanClient.update();
+            if (pricingTier != null && pricingTier.isPresent()) {
+                final com.azure.resourcemanager.appservice.models.PricingTier newPricingTier = AppServiceUtils.toPricingTier(pricingTier.get());
+                if (newPricingTier != AppServicePlan.this.getPlanService().pricingTier()) {
```
Comment (andxu) [inline azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkits/appservice/service/impl/AppServicePlan.java:155]: equals?

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkits/appservice/service/impl/WebAppDeploymentSlot.java [hunk 3]
```diff
@@ -34,81 +39,163 @@ public WebAppDeploymentSlot(WebAppDeploymentSlotEntity deploymentSlot, AzureAppS
...
+        return AppServiceUtils.getPublishingProfile(getDeploymentSlotClient().getPublishingProfile());
+    }
+
+    @Override
+    public WebAppDeploymentSlotEntity entity() {
+        return slotEntity;
+    }
+
+    private com.azure.resourcemanager.appservice.models.DeploymentSlot getDeploymentSlotClient() {
+        return getDeploymentSlotClient(false);
```
Comment (andxu) [inline azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkits/appservice/service/impl/WebAppDeploymentSlot.java:112]: I don't like the getXXX method with force parameter

Comment (andxu) [reply azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkits/appservice/service/impl/WebAppDeploymentSlot.java:112]: Look asc for a better name

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkits/appservice/service/impl/WebAppDeploymentSlot.java [hunk 4]
```diff
@@ -34,81 +39,163 @@ public WebAppDeploymentSlot(WebAppDeploymentSlotEntity deploymentSlot, AzureAppS
...
-        return null;
-    }
-}
\ No newline at end of file
+        return getDeploymentSlotClient().name();
+    }
+
+    @Getter
+    public class WebAppDeploymentSlotCreator implements IWebAppDeploymentSlotCreator {
+        public static final String CONFIGURATION_SOURCE_NEW = "new";
```
Comment (andxu) [inline azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkits/appservice/service/impl/WebAppDeploymentSlot.java:144]: priavte

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkits/appservice/service/impl/WebAppDeploymentSlot.java [hunk 5]
```diff
@@ -34,81 +39,163 @@ public WebAppDeploymentSlot(WebAppDeploymentSlotEntity deploymentSlot, AzureAppS
...
+                case CONFIGURATION_SOURCE_NEW:
+                    withCreate = blank.withBrandNewConfiguration();
+                    break;
+                case CONFIGURATION_SOURCE_PARENT:
+                    withCreate = blank.withConfigurationFromParent();
+                    break;
+                default:
+                    final DeploymentSlot deploymentSlot = deploymentSlotClient.parent().deploymentSlots().getByName(configurationSource);
+                    if (deploymentSlot == null) {
+                        throw new RuntimeException(CONFIGURATION_SOURCE_DOES_NOT_EXISTS);
```
Comment (andxu) [inline azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkits/appservice/service/impl/WebAppDeploymentSlot.java:188]: customize the error.

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkits/appservice/service/impl/WebAppDeploymentSlot.java [hunk 6]
```diff
@@ -34,81 +39,163 @@ public WebAppDeploymentSlot(WebAppDeploymentSlotEntity deploymentSlot, AzureAppS
...
+                    }
+                    withCreate = blank.withConfigurationFromDeploymentSlot(deploymentSlot);
+                    break;
+            }
+            if (appSettings != null && appSettings.isPresent()) {
+                withCreate.withAppSettings(appSettings.get());
+            }
+            WebAppDeploymentSlot.this.deploymentSlotClient = withCreate.create();
+            WebAppDeploymentSlot.this.slotEntity = AppServiceUtils.getWebAppDeploymentSlotEntity(WebAppDeploymentSlot.this.deploymentSlotClient);
```
Comment (andxu) [inline azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkits/appservice/service/impl/WebAppDeploymentSlot.java:197]: convertToWebAppDeploymentSlotEntity?

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkits/appservice/service/impl/AppServicePlan.java (hunk 1 above)
Comment (Flanker32) [reply azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkits/appservice/service/impl/AppServicePlan.java:88]: only name is shared by resources, for instance, region and resource group could not be set for `DeploymentSlotCreator`

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkits/appservice/service/impl/WebAppDeploymentSlot.java (hunk 4 above)
Comment (Flanker32) [reply azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkits/appservice/service/impl/WebAppDeploymentSlot.java:144]: They are designed to be public for tooling.

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkits/appservice/service/impl/WebAppDeploymentSlot.java (hunk 6 above)
Comment (Flanker32) [reply azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkits/appservice/service/impl/WebAppDeploymentSlot.java:199]: now convert to SDK models is all named `toXXX()`, convert to library model is all named `fromXXX()`, do we need to update the naming? As we discussed before, we could indicate the method from parameter type

Comment (andxu) [reply azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkits/appservice/service/impl/WebAppDeploymentSlot.java:199]: yes, toxx or fromxxx is ok, bug getXXX is not ok, it makes reader thinking that the entity is retrieved through api call to remote

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkits/appservice/service/impl/WebAppDeploymentSlot.java (hunk 3 above)
Comment (Flanker32) [reply azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkits/appservice/service/impl/WebAppDeploymentSlot.java]: change original force get method to refresh

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkits/appservice/service/impl/WebAppDeploymentSlot.java (hunk 6 above)
Comment (Flanker32) [reply azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkits/appservice/service/impl/WebAppDeploymentSlot.java]: change to `fromWebAppDeploymentSlot`

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkits/appservice/service/impl/WebApp.java [hunk 7]
```diff
@@ -194,7 +195,7 @@ public WebApp commit() {
                 withCreate.withAppSettings(getAppSettings().get());
             }
             WebApp.this.webAppClient = withCreate.create();
```
Comment (andxu) [inline azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkits/appservice/service/impl/WebApp.java:197]: webappInner

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkits/appservice/service/impl/WebApp.java [hunk 8]
```diff
@@ -194,7 +195,7 @@ public WebApp commit() {
...
-            WebApp.this.entity = AppServiceUtils.getWebAppEntity(WebApp.this.webAppClient);
+            WebApp.this.entity = AppServiceUtils.fromWebApp(WebApp.this.webAppClient);
```
Comment (andxu) [inline azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkits/appservice/service/impl/WebApp.java:198]: fromWebappInner

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkits/appservice/service/impl/WebApp.java [hunk 9]
```diff
@@ -138,22 +139,22 @@ public IWebAppDeploymentSlot deploymentSlot(String slotName) {
     }
 
     private com.azure.resourcemanager.appservice.models.WebApp getWebAppClient() {
```
Comment (andxu) [inline azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkits/appservice/service/impl/WebApp.java:141]: getWebappInner

**Review by andxu — APPROVED** [review]

**Review by andxu — APPROVED** [review]

**Pull request merged**